### PR TITLE
 Make yanking more compliant with bash conventions 

### PR DESCRIPTION
**Pull request by TsarFox**
In my comment on #10656 I mentioned that `kill_word` should move the killed region into the kill ring (emacs terminology, it's called "clipboard" in the code). I implemented this behavior for both `kill_word` and `kill_backwards_word`, as well as making the actual yanking more compliant (yanks to the current index, rather than to the end of the line).

Expect more tomorrow, as I'm planning to do more work to make Dietline more compliant with the readline conventions.

### CODE REVIEW

**Comment by XVilka** [comment]
You might want to check those issues too: https://github.com/radare/radare2/issues/9617 and https://github.com/radare/radare2/issues/9269

File: libr/cons/dietline.c [hunk 1]
```diff
@@ -1182,16 +1204,7 @@ R_API const char *r_line_readline_cb(RLineReadCallback cb, void *user) {
...
-				I.buffer.length += strlen (I.clipboard);
-				// TODO: support endless strings
-				if (I.buffer.length < R_LINE_BUFSIZE) {
-					I.buffer.index = I.buffer.length;
-					strcat (I.buffer.data, I.clipboard);
-				} else {
-					I.buffer.length -= strlen (I.clipboard);
-				}
-			}
+			yank();
```
Comment (radare) [inline libr/cons/dietline.c]: add missing space before (

Comment (radare) [reply libr/cons/dietline.c]: btw yank means copy, aka put buffer into clipboard. this case is doing a paste, not a yank

File: libr/cons/dietline.c [hunk 2]
```diff
@@ -68,11 +77,33 @@ static void kill_word() {
 	if (I.buffer.index >= I.buffer.length) {
 		I.buffer.length = I.buffer.index;
 	}
-	memmove (I.buffer.data + I.buffer.index, I.buffer.data + i,
-			I.buffer.length - i + 1);
+
+	len = i - I.buffer.index + 1;
+	free (I.clipboard);
+	I.clipboard = malloc (len + 1);
```
[the remark this answers is not in the document]
Comment (radare) [reply libr/cons/dietline.c]: err r_str_ndup

File: libr/cons/dietline.c [hunk 3]
```diff
@@ -50,6 +50,9 @@ static void backward_kill_word() {
 		if (I.buffer.index > I.buffer.length) {
 			I.buffer.length = I.buffer.index;
 		}
+		len = I.buffer.index - i + 1;
+		free (I.clipboard);
+		I.clipboard = r_str_ndup(I.buffer.data + i, len);
```
Comment (radare) [inline libr/cons/dietline.c]: space before (

File: libr/cons/dietline.c [hunk 4]
```diff
@@ -68,11 +71,27 @@ static void kill_word() {
 	if (I.buffer.index >= I.buffer.length) {
 		I.buffer.length = I.buffer.index;
 	}
-	memmove (I.buffer.data + I.buffer.index, I.buffer.data + i,
-			I.buffer.length - i + 1);
+	len = i - I.buffer.index + 1;
+	free (I.clipboard);
+	I.clipboard = r_str_ndup(I.buffer.data + I.buffer.index, len);
```
Comment (radare) [inline libr/cons/dietline.c]: space before (

File: libr/cons/dietline.c [hunk 5]
```diff
@@ -1182,16 +1192,7 @@ R_API const char *r_line_readline_cb(RLineReadCallback cb, void *user) {
...
-				I.buffer.length += strlen (I.clipboard);
-				// TODO: support endless strings
-				if (I.buffer.length < R_LINE_BUFSIZE) {
-					I.buffer.index = I.buffer.length;
-					strcat (I.buffer.data, I.clipboard);
-				} else {
-					I.buffer.length -= strlen (I.clipboard);
-				}
-			}
+			yank ();
```
Comment (radare) [inline libr/cons/dietline.c]: yank is not paste, bad naming?

File: libr/cons/dietline.c [hunk 6]
```diff
@@ -68,11 +71,27 @@ static void kill_word() {
...
 	I.buffer.length = strlen (I.buffer.data);
 }
 
+static void yank() {
+	int len, dist;
+	char *cursor;
+	if (I.clipboard) {
+		cursor = I.buffer.data + I.buffer.index;
+		dist = (I.buffer.data + I.buffer.length) - cursor;
+		len = strlen (I.clipboard);
```
Comment (radare) [inline libr/cons/dietline.c]: you can define and assign in here

File: libr/cons/dietline.c [hunk 7]
```diff
@@ -68,11 +71,27 @@ static void kill_word() {
...
+	free (I.clipboard);
+	I.clipboard = r_str_ndup(I.buffer.data + I.buffer.index, len);
+	memmove (I.buffer.data + I.buffer.index, I.buffer.data + i, len);
 	I.buffer.length = strlen (I.buffer.data);
 }
 
+static void yank() {
+	int len, dist;
+	char *cursor;
+	if (I.clipboard) {
```
Comment (radare) [inline libr/cons/dietline.c]: you can return earlier and avoid some nesting in here

File: libr/cons/dietline.c (hunk 1 above)
Comment (TsarFox) [reply libr/cons/dietline.c]: In emacs/readline terminology "yank" is equivalent to "paste". I think "yank" as copy is something vim did. I had the impression that the existing codebase was using the emacs/readline terminology per [this comment](https://github.com/radare/radare2/blob/eb7deb281df54771fb8ecf5890dc325a7d22d3e2/libr/cons/dietline.c#L712), but I would be happy to go back and change it if you would like.

File: libr/cons/dietline.c (hunk 5 above)
Comment (TsarFox) [reply libr/cons/dietline.c]: [See my comment on this here](https://github.com/radare/radare2/pull/10677#discussion_r200976323)

**Comment by radare** [comment]
https://stackoverflow.com/questions/16757516/why-yank-is-the-meaning-of-copy-in-vim <https://stackoverflow.com/questions/16757516/why-yank-is-the-meaning-of-copy-in-vim>

@Matt - In emacs you yank text from the kill ring and insert it into the document (a paste operation) whereas in vi you yank from the document into the paste buffer non destructively (a copyoperation). – Mark Booth <https://english.stackexchange.com/users/7231/mark-booth> Sep 5 '11 at 13:25 <https://english.stackexchange.com/questions/40657/how-yank-came-about-in-vi-and-emacs#comment74653_40657> 


so… in vim you have yank to copy and in emacs this is paste… great.. fuck emacs.

yanking means pulling something from one place. so i dont see how this can imply a paste. vim took this becayse ‘c’ was already taken to perform a copy.

File: libr/cons/dietline.c (hunk 7 above)
Comment (TsarFox) [reply libr/cons/dietline.c]: Do you mean checking for `I.clipboard` in the `case 25` statement rather than in the function?

**Comment by radare** [comment]
no i mean to if (!I.clipboard) {return} but thats just stylish. rename the function and ill merge

**Comment by TsarFox** [comment]
I renamed the function, but should I edit the comments that call `case 25` "yank" as well?

**Pull request merged**